File: src/entities/npc.js
// NPCs: hired employees who physically work the factory floor, and townsfolk who
// wander the world, trade and hand out contracts.

import { characterSheet, FW, FH } from '../art/characters.js';
import { clamp, dist, pick } from '../core/utils.js';
import { mulberry32 } from '../core/rng.js';

const FIRST = ['Mara', 'Tomas', 'Ines', 'Kwame', 'Yuki', 'Dario', 'Petra', 'Sven', 'Nadia', 'Bo',
  'Lucia', 'Otto', 'Rina', 'Ezra', 'Halle', 'Milo', 'Vera', 'Anders', 'Femi', 'Sasha',
  'Rosa', 'Hugo', 'Ilse', 'Jun', 'Karim', 'Elke', 'Nils', 'Tara', 'Ravi', 'Greta'];
const LAST = ['Vogel', 'Marek', 'Okafor', 'Lindqvist', 'Duarte', 'Novak', 'Haas', 'Bianchi',
  'Kowal', 'Rennes', 'Sato', 'Abara', 'Holt', 'Ferreira', 'Steen', 'Ivanov', 'Marchetti', 'Baptiste'];

export function randomName(rnd = Math.random) {
  return `${pick(FIRST, rnd)} ${pick(LAST, rnd)}`;
}

class Person {
  constructor(role, x, y, seed) {
    this.role = role; this.x = x; this.y = y; this.seed = seed;
    this.dir = 2; this.frame = 0; this.animT = 0;
    this.speed = 2.4 + (seed % 7) * 0.08;
    this.sheet = characterSheet(role, seed);
    this.target = null; this.path = null;
    this.state = 'idle'; this.stateT = 0;
    this.radius = 0.28;
  }

  stepToward(tx, ty, dt, game, speedMul = 1) {
    const dx = tx - this.x, dy = ty - this.y;
    const d = Math.hypot(dx, dy);
    if (d < 0.12) return true;
    const sp = this.speed * speedMul * dt;
    let nx = dx / d * sp, ny = dy / d * sp;
    if (Math.abs(nx) > Math.abs(ny)) this.dir = nx > 0 ? 1 : 3; else this.dir = ny > 0 ? 2 : 0;
    if (!game.blocked(this.x + nx, this.y, this.radius)) this.x += nx;
    else if (!game.blocked(this.x + nx, this.y + (ny > 0 ? 0.4 : -0.4), this.radius)) this.y += (ny > 0 ? 0.4 : -0.4) * 0.4;
    if (!game.blocked(this.x, this.y + ny, this.radius)) this.y += ny;
    else if (!game.blocked(this.x + (nx > 0 ? 0.4 : -0.4), this.y + ny, this.radius)) this.x += (nx > 0 ? 0.4 : -0.4) * 0.4;
    this.animT += dt * 8;
    this.frame = 1 + (Math.floor(this.animT) % 5);
    return false;
  }

  idleAnim(dt) { this.frame = 0; this.animT = 0; }

  draw(g, sx, sy) {
    g.drawImage(this.sheet.canvas, this.frame * FW, this.dir * FH, FW, FH, sx - FW / 2, sy - FH + 6, FW, FH);
  }
}

export class Employee extends Person {
  constructor(role, x, y, seed, name) {
    super(role, x, y, seed);
    const rnd = mulberry32(seed);
    this.name = name || randomName(rnd);
    this.skill = 1 + Math.floor(rnd() * 3);
    this.morale = 0.8;
    this.stamina = 1;
    this.wage = ROLE_WAGE[role] * (0.85 + this.skill * 0.12);
    this.assigned = null;
    this.carry = null;
    this.state = 'seek';
  }

  update(dt, game) {
    this.stateT -= dt;
    const f = game.factory;

    // rest when tired if there is a break room
    if (this.stamina < 0.15 && game.services.breakroomEnt) {
      this.state = 'rest'; this.assigned = null;
    }

    switch (this.state) {
      case 'rest': {
        const b = game.services.breakroomEnt;
        if (!b) { this.state = 'seek'; break; }
        if (this.stepToward(b.cx, b.cy, dt, game)) {
          this.idleAnim(dt);
          this.stamina = Math.min(1, this.stamina + dt * 0.12);
          this.morale = Math.min(1, this.morale + dt * 0.05);
          if (this.stamina >= 0.95) this.state = 'seek';
        }
        break;
      }
      case 'seek': {
        this.assigned = this.pickJob(game);
        this.state = this.assigned ? 'goto' : 'wander';
        this.stateT = 3;
        break;
      }
      case 'goto': {
        const t = this.assigned;
        if (!t || !f.ents.includes(t)) { this.state = 'seek'; break; }
        if (this.stepToward(t.cx, t.cy + 0.6, dt, game)) { this.state = 'work'; this.stateT = 6 + Math.random() * 6; t.workers++; }
        break;
      }
      case 'work': {
        const t = this.assigned;
        if (!t || !f.ents.includes(t)) { this.state = 'seek'; break; }
        this.frame = 6 + (Math.floor(this.animT * 2) % 2);
        this.animT += dt * 3;
        this.stamina = Math.max(0, this.stamina - dt * 0.012);
        // mechanics repair, everybody else boosts throughput
        if (this.role === 'mechanic' && (t.broken || t.condition < 0.9)) {
          f.repair(t, dt * 0.06 * this.skill * (game.services.maintenance ? 1.6 : 1));
          if (Math.random() < dt * 0.5) game.fx.sparks(t.cx, t.cy, 2);
        }
        if (this.stateT <= 0) { t.workers = Math.max(0, t.workers - 1); this.state = 'seek'; }
        break;
      }
      default: {   // wander inside the plot
        if (!this.wanderTarget || this.stateT <= 0) {
          const p = game.world.plot;
          this.wanderTarget = { x: p.x + 2 + Math.random() * (p.w - 4), y: p.y + 2 + Math.random() * (p.h - 4) };
          this.stateT = 5 + Math.random() * 5;
        }
        if (this.stepToward(this.wanderTarget.x, this.wanderTarget.y, dt, game, 0.6)) this.idleAnim(dt);
        if (this.stateT <= 0) this.state = 'seek';
        break;
      }
    }
    this.morale = clamp(this.morale - dt * 0.002 + (game.services.breakroom ? dt * 0.004 : 0), 0.2, 1);
  }

  pickJob(game) {
    const f = game.factory;
    let best = null, bestScore = -1;
    for (const e of f.ents) {
      let score = -1;
      if (this.role === 'mechanic') {
        if (e.broken) score = 100; else if (e.condition < 0.75) score = (1 - e.condition) * 40;
      } else if (this.role === 'logistics') {
        if (e.kind === 'store' || e.kind === 'dock') score = 12;
        else if (e.kind === 'machine' && e.totalOut() > 10) score = 18;
      } else if (this.role === 'researcher') {
        if (e.kind === 'lab') score = 40;
      } else if (this.role === 'engineer' || this.role === 'technician') {
        if (e.kind === 'machine' && e.active) score = 20 - e.workers * 6;
      } else {
        if ((e.kind === 'machine' || e.kind === 'miner') && !e.broken) score = 15 - e.workers * 5;
      }
      if (score <= 0) continue;
      score -= dist(this.x, this.y, e.cx, e.cy) * 0.25;
      if (score > bestScore) { bestScore = score; best = e; }
    }
    return best;
  }

  serialize() {
    return { r: this.role, x: this.x, y: this.y, s: this.seed, n: this.name,
      sk: this.skill, m: this.morale, st: this.stamina, w: this.wage };
  }
}

export const ROLE_WAGE = {
  worker: 22, logistics: 26, mechanic: 34, technician: 38, engineer: 52, researcher: 58, manager: 64,
};

export const ROLE_INFO = {
  worker:    { name: 'Factory Worker', desc: 'Operates machines, adding throughput wherever they stand.' },
  logistics: { name: 'Logistics Worker', desc: 'Keeps storage and shipping moving; faster dock turnaround.' },
  mechanic:  { name: 'Mechanic', desc: 'Walks the floor repairing worn and broken machines.' },
  technician:{ name: 'Technician', desc: 'Boosts machine efficiency and slows wear.' },
  engineer:  { name: 'Engineer', desc: 'Large throughput boost on complex machines.' },
  researcher:{ name: 'Researcher', desc: 'Staffs the lab, generating extra research points.' },
  manager:   { name: 'Manager', desc: 'Improves sale prices and lowers operating costs.' },
};

export class Townsfolk extends Person {
  constructor(x, y, seed, opts = {}) {
    super(opts.role || 'civilian', x, y, seed);
    const rnd = mulberry32(seed * 3 + 7);
    this.name = opts.name || randomName(rnd);
    this.home = { x, y };
    this.radiusRange = opts.range || 6;
    this.lines = opts.lines || [
      'The old foundry closed last winter. Plenty of hands looking for work.',
      'Hear the port pays well for finished goods.',
      'Careful in the mine — the deep seams flood after rain.',
      'Redhaven grows every year. Somebody has to build the machines.',
    ];
    this.vendor = !!opts.vendor;
    this.questGiver = !!opts.questGiver;
    this.title = opts.title || null;
  }

  update(dt, game) {
    this.stateT -= dt;
    if (this.stateT <= 0 || !this.wanderTarget) {
      this.wanderTarget = {
        x: this.home.x + (Math.random() - 0.5) * this.radiusRange * 2,
        y: this.home.y + (Math.random() - 0.5) * this.radiusRange * 2,
      };
      this.stateT = 4 + Math.random() * 6;
    }
    if (this.stepToward(this.wanderTarget.x, this.wanderTarget.y, dt, game, 0.55)) this.idleAnim(dt);
  }

  speak(rnd = Math.random) { return pick(this.lines, rnd); }
}

File: ironworks/entities/npc.py
"""NPCs: hired employees who physically work the factory floor, and townsfolk who
wander, trade and hand out contracts.
"""
import math

import pygame

from ..art.characters import character_sheet, FW, FH
from ..core.rng import Rng
from ..core.utils import clamp, dist

FIRST = ['Mara', 'Tomas', 'Ines', 'Kwame', 'Yuki', 'Dario', 'Petra', 'Sven', 'Nadia', 'Bo',
         'Lucia', 'Otto', 'Rina', 'Ezra', 'Halle', 'Milo', 'Vera', 'Anders', 'Femi', 'Sasha',
         'Rosa', 'Hugo', 'Ilse', 'Jun', 'Karim', 'Elke', 'Nils', 'Tara', 'Ravi', 'Greta']
LAST = ['Vogel', 'Marek', 'Okafor', 'Lindqvist', 'Duarte', 'Novak', 'Haas', 'Bianchi',
        'Kowal', 'Rennes', 'Sato', 'Abara', 'Holt', 'Ferreira', 'Steen', 'Ivanov',
        'Marchetti', 'Baptiste']

ROLE_WAGE = dict(worker=22, logistics=26, mechanic=34, technician=38,
                 engineer=52, researcher=58, manager=64)

ROLE_INFO = {
    'worker':     dict(name='Factory Worker', desc='Operates machines, adding throughput wherever they stand.'),
    'logistics':  dict(name='Logistics Worker', desc='Keeps storage and shipping moving; faster dock turnaround.'),
    'mechanic':   dict(name='Mechanic', desc='Walks the floor repairing worn and broken machines.'),
    'technician': dict(name='Technician', desc='Boosts machine efficiency and slows wear.'),
    'engineer':   dict(name='Engineer', desc='Large throughput boost on complex machines.'),
    'researcher': dict(name='Researcher', desc='Staffs the lab, generating extra research points.'),
    'manager':    dict(name='Manager', desc='Improves sale prices and lowers operating costs.'),
}


def random_name(rnd):
    return '%s %s' % (rnd.pick(FIRST), rnd.pick(LAST))


class Person:
    def __init__(self, role, x, y, seed):
        self.role = role
        self.x = float(x)
        self.y = float(y)
        self.seed = int(seed)
        self.dir = 2
        self.frame = 0
        self.anim_t = 0.0
        self.speed = 2.4 + (self.seed % 7) * 0.08
        self.sheet = character_sheet(role, self.seed)
        self.state = 'idle'
        self.state_t = 0.0
        self.radius = 0.28
        self.wander_target = None

    def step_toward(self, tx, ty, dt, game, speed_mul=1.0):
        dx, dy = tx - self.x, ty - self.y
        d = math.hypot(dx, dy)
        if d < 0.12:
            return True
        sp = self.speed * speed_mul * dt
        nx, ny = dx / d * sp, dy / d * sp
        self.dir = (1 if nx > 0 else 3) if abs(nx) > abs(ny) else (2 if ny > 0 else 0)
        if not game.blocked(self.x + nx, self.y, self.radius):
            self.x += nx
        elif not game.blocked(self.x + nx, self.y + (0.4 if ny > 0 else -0.4), self.radius):
            self.y += (0.4 if ny > 0 else -0.4) * 0.4
        if not game.blocked(self.x, self.y + ny, self.radius):
            self.y += ny
        elif not game.blocked(self.x + (0.4 if nx > 0 else -0.4), self.y + ny, self.radius):
            self.x += (0.4 if nx > 0 else -0.4) * 0.4
        self.anim_t += dt * 8
        self.frame = 1 + int(self.anim_t) % 5
        return False

    def idle(self):
        self.frame = 0
        self.anim_t = 0.0

    def draw(self, target, sx, sy):
        src = pygame.Rect(self.frame * FW, self.dir * FH, FW, FH)
        target.blit(self.sheet['surf'], (int(sx - FW / 2), int(sy - FH + 6)), src)


class Employee(Person):
    def __init__(self, role, x, y, seed, name=None):
        super().__init__(role, x, y, seed)
        rnd = Rng(self.seed)
        self.name = name or random_name(rnd)
        self.skill = 1 + int(rnd() * 3)
        self.morale = 0.8
        self.stamina = 1.0
        self.wage = ROLE_WAGE[role] * (0.85 + self.skill * 0.12)
        self.assigned = None
        self.state = 'seek'

    def update(self, dt, game):
        self.state_t -= dt
        factory = game.factory

        if self.stamina < 0.15 and game.services.get('breakroom_ent'):
            self.state = 'rest'
            self.assigned = None

        if self.state == 'rest':
            b = game.services.get('breakroom_ent')
            if not b:
                self.state = 'seek'
            elif self.step_toward(b.cx, b.cy, dt, game):
                self.idle()
                self.stamina = min(1.0, self.stamina + dt * 0.12)
                self.morale = min(1.0, self.morale + dt * 0.05)
                if self.stamina >= 0.95:
                    self.state = 'seek'
        elif self.state == 'seek':
            self.assigned = self.pick_job(game)
            self.state = 'goto' if self.assigned else 'wander'
            self.state_t = 3
        elif self.state == 'goto':
            t = self.assigned
            if t is None or t not in factory.ents:
                self.state = 'seek'
            elif self.step_toward(t.cx, t.cy + 0.6, dt, game):
                self.state = 'work'
                self.state_t = 6 + (self.seed % 6)
                t.workers += 1
        elif self.state == 'work':
            t = self.assigned
            if t is None or t not in factory.ents:
                self.state = 'seek'
            else:
                self.anim_t += dt * 3
                self.frame = 6 + int(self.anim_t * 2) % 2
                self.stamina = max(0.0, self.stamina - dt * 0.012)
                if self.role == 'mechanic' and (t.broken or t.condition < 0.9):
                    factory.repair(t, dt * 0.06 * self.skill *
                                   (1.6 if game.services.get('maintenance') else 1.0))
                    if game.rnd() < dt * 0.5:
                        game.fx.sparks(t.cx, t.cy, 2)
                if self.state_t <= 0:
                    t.workers = max(0, t.workers - 1)
                    self.state = 'seek'
        else:
            if self.wander_target is None or self.state_t <= 0:
                p = game.world.plot
                self.wander_target = (p['x'] + 2 + game.rnd() * (p['w'] - 4),
                                      p['y'] + 2 + game.rnd() * (p['h'] - 4))
                self.state_t = 5 + game.rnd() * 5
            if self.step_toward(self.wander_target[0], self.wander_target[1], dt, game, 0.6):
                self.idle()
            if self.state_t <= 0:
                self.state = 'seek'

        self.morale = clamp(self.morale - dt * 0.002 +
                            (dt * 0.004 if game.services.get('breakroom') else 0), 0.2, 1.0)

    def pick_job(self, game):
        best, best_score = None, -1
        for e in game.factory.ents:
            score = -1
            if self.role == 'mechanic':
                if e.broken:
                    score = 100
                elif e.condition < 0.75:
                    score = (1 - e.condition) * 40
            elif self.role == 'logistics':
                if e.kind in ('store', 'dock'):
                    score = 12
                elif e.kind == 'machine' and e.total_out() > 10:
                    score = 18
            elif self.role == 'researcher':
                if e.kind == 'lab':
                    score = 40
            elif self.role in ('engineer', 'technician'):
                if e.kind == 'machine' and e.active:
                    score = 20 - e.workers * 6
            else:
                if e.kind in ('machine', 'miner') and not e.broken:
                    score = 15 - e.workers * 5
            if score <= 0:
                continue
            score -= dist(self.x, self.y, e.cx, e.cy) * 0.25
            if score > best_score:
                best_score, best = score, e
        return best

    def serialize(self):
        return dict(r=self.role, x=self.x, y=self.y, s=self.seed, n=self.name,
                    sk=self.skill, m=self.morale, st=self.stamina, w=self.wage)


class Townsfolk(Person):
    def __init__(self, x, y, seed, role='civilian', name=None, title=None,
                 lines=None, range_=6, vendor=False, quest_giver=False):
        super().__init__(role, x, y, seed)
        rnd = Rng(self.seed * 3 + 7)
        self.name = name or random_name(rnd)
        self.home = (x, y)
        self.range = range_
        self.title = title
        self.vendor = vendor
        self.quest_giver = quest_giver
        self.lines = lines or [
            'The old foundry closed last winter. Plenty of hands looking for work.',
            'Hear the port pays well for finished goods.',
            'Careful in the mine - the deep seams flood after rain.',
            'Redhaven grows every year. Somebody has to build the machines.',
        ]
        self._line = 0

    def update(self, dt, game):
        self.state_t -= dt
        if self.state_t <= 0 or self.wander_target is None:
            self.wander_target = (self.home[0] + (game.rnd() - 0.5) * self.range * 2,
                                  self.home[1] + (game.rnd() - 0.5) * self.range * 2)
            self.state_t = 4 + game.rnd() * 6
        if self.step_toward(self.wander_target[0], self.wander_target[1], dt, game, 0.55):
            self.idle()

    def speak(self):
        line = self.lines[self._line % len(self.lines)]
        self._line += 1
        return line
